src/rigid_pendulum_poc/solver.cpp

```diff
@@ -57,4 +54,190 @@ void solve_linear_system(
+        // TODO: Provide actual force increments
+        auto increments = residuals;
+
+        if (this->CheckConvergence(residuals, increments)) {
+            break;
+        }
+
+        auto iteration_matrix = ComputeIterationMatrix(gen_coords);
+        auto gen_coords_delta = residuals;
+        solve_linear_system(iteration_matrix, gen_coords_delta);
+
+        Kokkos::parallel_for(
+            size,
+            KOKKOS_LAMBDA(const int i) {
+                gen_coords(i) += gen_coords_delta(i);
+                gen_velocity(i) += gamma_prime * gen_coords_delta(i);
+                gen_accln(i) += beta_prime * gen_coords_delta(i);
+            }
+        );
+    }
+
+    log->Debug("Converged in " + std::to_string(this->n_iterations_) + " iterations\n");
+    this->total_n_iterations_ += this->n_iterations_;
+
+    Kokkos::parallel_for(
+        size,
+        KOKKOS_LAMBDA(const int i) {
+            // Update algorithmic acceleration once soln has converged
+            algo_accln(i) += (1 - kALPHA_F) / (1 - kALPHA_M) * gen_accln(i);
+        }
+    );
+
+    return State(gen_coords, gen_velocity, gen_accln, algo_accln);
+}
+
+std::tuple<HostView1D, HostView1D, HostView1D> GeneralizedAlphaTimeIntegrator::UpdateLinearSolution(
+    HostView1D gen_coords, HostView1D gen_velocity, HostView1D gen_accln, HostView1D algo_accln
+) {
+    // Update generalized coordinates, generalized velocity, and algorithmic acceleration
+    // based on generalized coordinates, generalized velocity, generalized acceleration,
+    // and algorithmic acceleration from previous time step and algorithmic acceleration
+    // from current time step
+    const auto h = this->time_step_;
+    const auto size = gen_coords.size();
+
+    Kokkos::parallel_for(
+        size,
+        KOKKOS_LAMBDA(const int i) {
+            gen_coords(i) += h * gen_velocity(i) + h * h * (0.5 - kBETA) * algo_accln(i);
+            gen_velocity(i) += h * (1 - kGAMMA) * algo_accln(i);
+            algo_accln(i) =
+                (1.0 / (1.0 - kALPHA_M)) * (kALPHA_F * gen_accln(i) - kALPHA_M * algo_accln(i));
+            gen_coords(i) += h * h * kBETA * algo_accln(i);
+            gen_velocity(i) += h * kBETA * algo_accln(i);
+        }
+    );
+
+    auto log = util::Log::Get();
+    log->Debug("Linear solution: gen_coords, gen_velocity, algo_acceleration\n");
+
+    for (size_t i = 0; i < size; i++) {
+        log->Debug(
+            "row " + std::to_string(i) + ": " + std::to_string(gen_coords(i)) + "\t" +
+            std::to_string(gen_velocity(i)) + "\t" + std::to_string(algo_accln(i)) + "\n"
+        );
+    }
+
+    return {gen_coords, gen_velocity, algo_accln};
+}
+
+HostView1D GeneralizedAlphaTimeIntegrator::ComputeResiduals(HostView1D forces) {
+    // TODO: Compute the residuals
+    // r^q = M(q) * q'' - f + phi_q^T * lambda
+    // This is a just a placeholder, returns the provided forces as the residuals
+    return forces;
```

Comment:
If we return a vector of all ones here (of the same size as forces), it should make testing really simple.  

Once this testing is in place, you can abstract this functionality out to an object that has its own "getResidual" and "getInterationMatrix" functions.

Reply:
> If we return a vector of all ones here (of the same size as forces), it should make testing really simple.

This sounds like a reasonable approach to me, with following caveat.

We are always going to have either convergence/non-convergence ([the check at line 129 here](https://github.com/Exawind/openturbine/blob/715156ff17465c8dd966db176275cbda81d83da6/src/rigid_pendulum_poc/solver.cpp#L129)) for the `AlphaStep` based on the hard-coded behavior here. That would prevent us from observing the convergence of the `AlphaStep` after, let's say, one iteration. Any thoughts on how to test out the behavior of the GA solver without implementing the details of the `ComputeResiduals()` function?

FWIW, my opinion is to return non-convergence based on the hard-coded residual and iteration matrix and test the expected solution after one iteration of the GA time integration by setting the [max_iterations](https://github.com/Exawind/openturbine/blob/715156ff17465c8dd966db176275cbda81d83da6/src/rigid_pendulum_poc/solver.h#L52) at 1. What do you think?

Reply:
I think the first step is to use a simple system that will converge after one nonlinear iteration.  This is easy to both formulate and test.  I'd first start with an identity iteration matrix like you have.  Once that's working, you can maybe put together a non-singular, non-identity iteration matrix that you know the exact solution for to ensure everything is working for a matrix that isn't just the identity.  The same logic follows for the residual - start with something simple like all ones, then we can do a static residual vector that is a little more complex.

The nonlinear case is a little harder to implement the residual/iteration matrix.  Once you have those, I agree with setting max_iterations to 1 to test the behavior (since you can easily compute the result by hand).  I'd even set max_iterations to 2 to test the values after that to compare to a hand computed value (assuming that the system is chosen so that it is possible).  Once you know it works to 2 iterations, it should work for all greater numbers.  

The convergence check itself might be harder to test, but we can cross that bridge when we get to it.